File: scripts/backfill_claims.py
#!/usr/bin/env python3
"""Claim Ledger Stage 2 — backfill approved claims from entity rows (Python).

Usage:
  python scripts/backfill_claims.py --all-sites
  python scripts/backfill_claims.py --site-id <uuid>
  python scripts/backfill_claims.py --all-sites --dry-run
  python scripts/backfill_claims.py --all-sites --include-pending
  python scripts/backfill_claims.py --all-sites --update   # supersede on value change

Idempotency (default): skip when an approved claim already exists with the same
value for (entity_id, attribute). With --update, value changes insert a new
approved claim and mark the previous approved as status=superseded.
"""

from __future__ import annotations

import argparse
import sys
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from sqlalchemy import select  # noqa: E402
from sqlalchemy.orm import Session  # noqa: E402

from app.db import SessionLocal, init_db  # noqa: E402
from app.models import Claim, Entity  # noqa: E402
from app.services.claim_codec import entity_attribute_pairs  # noqa: E402


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _latest_approved(db: Session, entity_id: str, attribute: str) -> Claim | None:
    rows = list(
        db.scalars(
            select(Claim)
            .where(
                Claim.entity_id == entity_id,
                Claim.attribute == attribute,
                Claim.status == "approved",
            )
            .order_by(Claim.id.desc())
        ).all()
    )
    return rows[0] if rows else None


def backfill_entity(
    db: Session,
    entity: Entity,
    *,
    dry_run: bool = False,
    update: bool = False,
) -> dict[str, int]:
    """Backfill claims for one entity. Returns counters."""
    stats = {"inserted": 0, "skipped": 0, "superseded": 0}
    pairs = entity_attribute_pairs(entity)
    approved_at = entity.last_updated or entity.created_at or _utcnow()
    extraction = (entity.source or "scan")[:32]
    entity_type = entity.entity_type or "unknown"

    for attr, value in pairs:
        existing = _latest_approved(db, entity.id, attr)
        if existing is not None:
            if existing.value == value:
                stats["skipped"] += 1
                continue
            if not update:
                stats["skipped"] += 1
                continue
            # Supersede previous approved
            if not dry_run:
                existing.status = "superseded"
                claim = Claim(
                    entity_id=entity.id,
                    entity_type=entity_type,
                    attribute=attr,
                    value=value,
                    source_url=None,
                    extraction_method=extraction,
                    confidence=None,
                    status="approved",
                    supersedes_id=existing.id,
                    created_at=approved_at,
                    approved_by="backfill",
                    approved_at=approved_at,
                )
                db.add(claim)
            stats["superseded"] += 1
            stats["inserted"] += 1
            continue

        if not dry_run:
            claim = Claim(
                entity_id=entity.id,
                entity_type=entity_type,
                attribute=attr,
                value=value,
                source_url=None,
                extraction_method=extraction,
                confidence=None,
                status="approved",
                supersedes_id=None,
                created_at=approved_at,
                approved_by="backfill",
                approved_at=approved_at,
            )
            db.add(claim)
        stats["inserted"] += 1

    return stats


def backfill_site(
    db: Session,
    site_id: str | None,
    *,
    include_pending: bool = False,
    dry_run: bool = False,
    update: bool = False,
) -> dict[str, int]:
    q = select(Entity)
    if site_id:
        q = q.where(Entity.site_id == site_id)
    if include_pending:
        q = q.where(Entity.status.in_(["approved", "pending", "needs_edit"]))
    else:
        q = q.where(Entity.status == "approved")

    entities = list(db.scalars(q.order_by(Entity.entity_type, Entity.name)).all())
    totals = {
        "entities": 0,
        "inserted": 0,
        "skipped": 0,
        "superseded": 0,
    }
    for ent in entities:
        totals["entities"] += 1
        s = backfill_entity(db, ent, dry_run=dry_run, update=update)
        for k in ("inserted", "skipped", "superseded"):
            totals[k] += s[k]
    if not dry_run:
        db.commit()
    return totals


def main() -> int:
    parser = argparse.ArgumentParser(description="Backfill claims from entity rows (Stage 2)")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--site-id", help="Backfill one site")
    group.add_argument("--all-sites", action="store_true", help="Backfill all sites")
    parser.add_argument(
        "--include-pending",
        action="store_true",
        help="Also backfill pending/needs_edit entities (default: approved only)",
    )
    parser.add_argument("--dry-run", action="store_true", help="Count only; no writes")
    parser.add_argument(
        "--update",
        action="store_true",
        help="On value change, supersede previous approved claim",
    )
    args = parser.parse_args()

    init_db()
    db = SessionLocal()
    try:
        site_id = None if args.all_sites else args.site_id
        totals = backfill_site(
            db,
            site_id,
            include_pending=args.include_pending,
            dry_run=args.dry_run,
            update=args.update,
        )
        mode = "DRY-RUN " if args.dry_run else ""
        print(
            f"{mode}Stage 2 backfill: entities={totals['entities']} "
            f"inserted={totals['inserted']} skipped={totals['skipped']} "
            f"superseded={totals['superseded']}"
        )
        return 0
    finally:
        db.close()


if __name__ == "__main__":
    raise SystemExit(main())

File: php/scripts/backfill_claims.php
<?php
declare(strict_types=1);

/**
 * Claim Ledger Stage 2 — backfill approved claims from entity rows (PHP Host).
 *
 * Usage:
 *   php php/scripts/backfill_claims.php --db=/path/to/bkbs.sqlite
 *   php php/scripts/backfill_claims.php --db=... --site-id=<uuid>
 *   php php/scripts/backfill_claims.php --db=... --dry-run
 *   php php/scripts/backfill_claims.php --db=... --include-pending
 *   php php/scripts/backfill_claims.php --db=... --update
 *
 * Default: approved entities only; skip identical existing approved claims.
 */

// Packaged zip: bkbs-php/scripts → bkbs-php/src
// Monorepo: php/scripts → php/src (or repo/php/src)
$bootstrap = null;
foreach (
    [
        dirname(__DIR__) . '/src/bootstrap.php',
        dirname(__DIR__, 2) . '/php/src/bootstrap.php',
        dirname(__DIR__) . '/bootstrap.php',
    ] as $cand
) {
    if (is_file($cand)) {
        $bootstrap = $cand;
        break;
    }
}
if ($bootstrap === null) {
    fwrite(STDERR, "Cannot locate php/src/bootstrap.php from " . __DIR__ . "\n");
    exit(1);
}
require $bootstrap;

use Bkbs\Database;
use Bkbs\Resolver;

$opts = getopt('', ['db:', 'site-id:', 'all-sites', 'dry-run', 'include-pending', 'update', 'help']);
if (isset($opts['help']) || empty($opts['db'])) {
    fwrite(STDOUT, "Usage: php php/scripts/backfill_claims.php --db=PATH [--site-id=UUID] [--all-sites] [--dry-run] [--include-pending] [--update]\n");
    exit(isset($opts['help']) ? 0 : 1);
}

$dbPath = (string) $opts['db'];
$siteId = isset($opts['site-id']) ? (string) $opts['site-id'] : null;
$dryRun = isset($opts['dry-run']);
$includePending = isset($opts['include-pending']);
$update = isset($opts['update']);

$db = new Database($dbPath);
$pdo = $db->pdo();

$sql = 'SELECT * FROM entities';
$params = [];
$where = [];
if ($siteId !== null && $siteId !== '') {
    $where[] = 'site_id = ?';
    $params[] = $siteId;
}
if ($includePending) {
    $where[] = "status IN ('approved','pending','needs_edit')";
} else {
    $where[] = "status = 'approved'";
}
if ($where) {
    $sql .= ' WHERE ' . implode(' AND ', $where);
}
$sql .= ' ORDER BY entity_type, name';

$st = $pdo->prepare($sql);
$st->execute($params);
$entities = $st->fetchAll(PDO::FETCH_ASSOC);

$totals = ['entities' => 0, 'inserted' => 0, 'skipped' => 0, 'superseded' => 0];

$findApproved = $pdo->prepare(
    'SELECT id, value FROM claims WHERE entity_id = ? AND attribute = ? AND status = ? ORDER BY id DESC LIMIT 1'
);
$insert = $pdo->prepare(
    'INSERT INTO claims(entity_id, entity_type, attribute, value, source_url, extraction_method, confidence, status, supersedes_id, created_at, approved_by, approved_at, review_due_at)
     VALUES(?,?,?,?,NULL,?,NULL,?,?,?,?,?,NULL)'
);
$supersede = $pdo->prepare('UPDATE claims SET status = ? WHERE id = ?');

if (!$dryRun) {
    $pdo->beginTransaction();
}

foreach ($entities as $ent) {
    $totals['entities']++;
    $entityId = (string) $ent['id'];
    $entityType = (string) ($ent['entity_type'] ?? 'unknown');
    $extraction = substr((string) ($ent['source'] ?? 'scan'), 0, 32);
    $approvedAt = (string) ($ent['last_updated'] ?? $ent['created_at'] ?? gmdate('c'));
    $pairs = Resolver::entityAttributePairs($ent);

    foreach ($pairs as [$attr, $value]) {
        $findApproved->execute([$entityId, $attr, 'approved']);
        $existing = $findApproved->fetch(PDO::FETCH_ASSOC);
        if ($existing) {
            if ((string) $existing['value'] === $value) {
                $totals['skipped']++;
                continue;
            }
            if (!$update) {
                $totals['skipped']++;
                continue;
            }
            if (!$dryRun) {
                $supersede->execute(['superseded', (int) $existing['id']]);
                $insert->execute([
                    $entityId,
                    $entityType,
                    $attr,
                    $value,
                    $extraction,
                    'approved',
                    (int) $existing['id'],
                    $approvedAt,
                    'backfill',
                    $approvedAt,
                ]);
            }
            $totals['superseded']++;
            $totals['inserted']++;
            continue;
        }
        if (!$dryRun) {
            $insert->execute([
                $entityId,
                $entityType,
                $attr,
                $value,
                $extraction,
                'approved',
                null,
                $approvedAt,
                'backfill',
                $approvedAt,
            ]);
        }
        $totals['inserted']++;
    }
}

if (!$dryRun) {
    $pdo->commit();
}

$mode = $dryRun ? 'DRY-RUN ' : '';
echo "{$mode}Stage 2 backfill (PHP): entities={$totals['entities']} inserted={$totals['inserted']} skipped={$totals['skipped']} superseded={$totals['superseded']}\n";
exit(0);
